From a35138d49e022dd94fd91974fa2ed2ef6dd58a90 Mon Sep 17 00:00:00 2001
From: Simon McVittie <smcv@debian.org>
Date: Sat, 3 Jan 2009 12:04:48 +0000
Subject: [PATCH 11/13] Add new #defines to indicate which syscall is umount(3) and which is umount2(3).

* on architectures where __NR_umount is umount(3) and __NR_umount2 is
  umount2(3), don't do anything special
* on architectures where this is not the case, define __NR_umount_with_flags
  so that it's umount2(3)
* define __NR_umount_without_flags to be umount(3) on architectures where
  such a syscall exists

In the currently-supported architectures there are four families:

* on i386, arm etc., __NR_umount takes one argument and __NR_umount2 takes two
* on x86_64 and parisc __NR_umount2 takes two arguments and there is no
  1-argument umount
* on alpha, __NR_oldumount takes one argument and __NR_umount takes two
* on ia64, __NR_umount takes two arguments and there is no 1-argument umount
---
 alpha/syscalls.h     |    2 ++
 ia64/syscalls.h      |    2 ++
 syscalls.s/umount.S  |    4 ++++
 syscalls.s/umount2.S |    4 +++-
 4 files changed, 11 insertions(+), 1 deletions(-)

--- a/alpha/syscalls.h
+++ b/alpha/syscalls.h
@@ -231,6 +231,8 @@
 #define __NR_osf_memcntl	260	/* not implemented */
 #define __NR_osf_fdatasync	261	/* not implemented */
 
+#define __NR_umount_without_flags __NR_oldumount
+#define __NR_umount_with_flags __NR_umount
 
 /*
  * Linux-specific system calls begin at 300
--- a/ia64/syscalls.h
+++ b/ia64/syscalls.h
@@ -318,6 +318,8 @@
 #define __NR_getrandom			1339
 #define __NR_memfd_create		1340
 
+#define __NR_umount_with_flags __NR_umount
+
 #define syscall(name, sym) \
 .text; \
 .globl sym; \
--- a/syscalls.s/umount.S
+++ b/syscalls.s/umount.S
@@ -1,5 +1,9 @@
 #include "syscalls.h"
 
 #ifdef __NR_umount
+#if defined(__NR_umount_without_flags)
+syscall(umount_without_flags,umount)
+#elif !defined(__NR_umount_with_flags) || (__NR_umount != __NR_umount_with_flags)
 syscall(umount,umount)
 #endif
+#endif
--- a/syscalls.s/umount2.S
+++ b/syscalls.s/umount2.S
@@ -1,5 +1,7 @@
 #include "syscalls.h"
 
-#ifdef __NR_umount2
+#if defined(__NR_umount_with_flags)
+syscall(umount_with_flags,umount2)
+#elif defined(__NR_umount2)
 syscall(umount2,umount2)
 #endif
